contracts/profiles/StealthKeyResolver.sol

```diff
@@ -4,36 +4,83 @@ import "../ResolverBase.sol";
 
-    struct StealthKey {
-        uint256 generationPubKey;
-        uint256 encryptionPubKey;
-    }
-
-    mapping(bytes32=>StealthKey) _stealthKeys;
+    /**
+     * @dev Mapping used to store two secp256k1 curve public keys useful for
+     * receiving stealth payments. The mapping records two keys: a viewing
+     * key and a spending key, which can be set and read via the `setsStealthKeys`
+     * and `stealthKey` methods respectively.
+     *
+     * The mapping associates the node to another mapping, which itself maps
+     * the public key prefix to the actual key . This scheme is used to avoid using an
+     * extra storage slot for the public key prefix. For a given node, the mapping
+     * may contain a spending key at position 0 or 1, and a viewing key at position
+     * 2 or 3. See the setter/getter methods for details of how these map to prefixes.
+     *
+     * For more on secp256k1 public keys and prefixes generally, see:
+     * https://github.com/ethereumbook/ethereumbook/blob/develop/04keys-addresses.asciidoc#generating-a-public-key
+     *
+     */
+    mapping(bytes32 => mapping(uint256 => uint256)) _stealthKeys;
 
     /**
      * Sets the stealth keys associated with an ENS name, for anonymous sends.
      * May only be called by the owner of that node in the ENS registry.
      * @param node The node to update.
-     * @param generationPubKey The public key for generating a stealth address
-     * @param encryptionPubKey The public key to use for encryption
+     * @param spendingPubKeyPrefix Prefix of the spending public key (2 or 3)
+     * @param spendingPubKey The public key for generating a stealth address
+     * @param viewingPubKeyPrefix Prefix of the viewing public key (2 or 3)
+     * @param viewingPubKey The public key to use for encryption
      */
-    function setStealthKeys(bytes32 node, uint256 generationPubKey, uint256 encryptionPubKey) external authorised(node) {
-        _stealthKeys[node] = StealthKey(generationPubKey, encryptionPubKey);
-        emit StealthKeyChanged(node, generationPubKey, encryptionPubKey);
+    function setStealthKeys(bytes32 node, uint256 spendingPubKeyPrefix, uint256 spendingPubKey, uint256 viewingPubKeyPrefix, uint256 viewingPubKey) external authorised(node) {
+        require(
+            (spendingPubKeyPrefix == 2 || spendingPubKeyPrefix == 3) &&
+            (viewingPubKeyPrefix == 2 || viewingPubKeyPrefix == 3),
+            "StealthKeyResolver: Invalid Prefix"
+        );
+
+        emit StealthKeyChanged(node, spendingPubKeyPrefix, spendingPubKey, viewingPubKeyPrefix, viewingPubKey);
+
+        // Shift the spending key prefix down by 2, making it the appropriate index of 0 or 1
+        spendingPubKeyPrefix -= 2;
+
+        // Ensure the opposite prefix indices are empty
+        delete _stealthKeys[node][spendingPubKeyPrefix == 0 ? 1 : 0];
+        delete _stealthKeys[node][viewingPubKeyPrefix == 2 ? 3 : 2];
```

Comment:
```suggestion
        delete _stealthKeys[node][1 - spendingPubKeyPrefix];
        delete _stealthKeys[node][3 - viewingPubKeyPrefix];
```

Reply:
Very clever! One minor error here, for line 51 the prefix needs to be subtracted from 5 for it to work, since we want it to resolve to 2 or 3. I pushed a new commit with this fix. Let me know if see anything else!